Mixpanel/MPABTestDesignerConnection.m

```diff
@@ -218,21 +222,24 @@ - (void)webSocket:(MPWebSocket *)webSocket didCloseWithCode:(NSInteger)code reas
     _open = NO;
     if (_connected) {
         _connected = NO;
-        [self reconnect:YES];
+        [self reconnect:YES maxInterval:10 maxRetries:10];
         if (_disconnectCallback) {
             _disconnectCallback();
         }
+    } else if (_keepTrying) {
+        [self reconnect:YES maxInterval:30 maxRetries:40];
     }
 }
 
-- (void)reconnect:(BOOL)initiate
+- (void)reconnect:(BOOL)initiate maxInterval:(int)maxInterval maxRetries:(int)maxRetries
 {
     static int retries = 0;
-    if (self.sessionEnded || _connected || retries >= 10) {
+    
+    if (self.sessionEnded || _connected || retries >= maxRetries) {
         // If we deliberately closed the connection, or are already connected
         // or we tried too many times, then stop retrying.
         retries = 0;
-    } else if(initiate ^ (retries > 0)) {
+    } else if (initiate ^ (retries > 0)) {
```

Comment:
@joeatwork  great question. yep, that's a XOR, the reasoning is this:

There are 3 cases.
`initiate = YES && (retries > 0) == NO`: This means you were already happily using the socket connection and it broke because the socket died somehow. That's when we try and get reconnected by calling`[socket open]` and also throwing a retry into the future in case this `[socket open]` doesn't work.

`initiate = NO && (retries > 0) == YES`: This is when we are in an exponential backoff retry loop, and somehow the first `[socket open]` didn't work. (ie `_connected` is still NO), so let's try again. (and kick off another retry even further in the future in case this one doesn't work either. 

`initiate = YES && (retries > 0) == YES`: This is the case where our `[socket open]` kindof worked but didn't go well (probably because the switchboard immediately rejected the connection for some reason) and the `webSocket:didFailWithError:` or `webSocket:didCloseWithCode` handler calls `reconnect` again because of that. In this case we are already in a retry backoff loop, so we don't actually want to start another one (because we know there is a queued up call to `reconnect` already scheduled for the future, or the maximum number of retries has already been reached) 

So yeah, TL;DR, we want to retry if we are either starting a reconnect cycle, or if we are already in one, but we don't want to start another cycle if we are already in one.